### PR TITLE
Benchmarks - Support TE FP8 in BERT/GPT2 models

### PR DESCRIPTION
**Pull request by abuccts**
Support Transformer Engine FP8 in existing PyTorch BERT/GPT2 models by converting linear/layernorm to TE layers.

### CODE REVIEW

**Review by cp5555 — APPROVED** [review]

File: superbench/benchmarks/model_benchmarks/pytorch_bert.py [hunk 1]
```diff
@@ -183,15 +183,15 @@ def _create_model(self, precision):
...
         try:
+            self._model = BertBenchmarkModel(self._config, self._args.num_classes)
             if enable_fp8:
                 self._fp8_recipe = DelayedScaling(
                     fp8_format=Format[precision.name.strip('FP8_')],
                     amax_history_len=16,
                     amax_compute_algo='max',
                 )
-                self._model = TeBertBenchmarkModel(self._config, self._args.num_classes).to(dtype=torch.float16)
+                self._to_te_model(self._model.to(dtype=torch.float16))
```
Comment (cp5555) [inline superbench/benchmarks/model_benchmarks/pytorch_bert.py:193]: Can we remove whole TeBertBenchmarkModel class?

File: superbench/benchmarks/model_benchmarks/pytorch_gpt2.py [hunk 2]
```diff
@@ -53,7 +58,13 @@ def __init__(self, name, parameters=''):
         """
         super().__init__(name, parameters)
         self._config = None
-        self._supported_precision = [Precision.FLOAT32, Precision.FLOAT16]
+        self._fp8_recipe = None
+        self._supported_precision = [
+            Precision.FLOAT32,
+            Precision.FLOAT16,
+            Precision.FP8_HYBRID,
+            Precision.FP8_E4M3,
```
Comment (cp5555) [inline superbench/benchmarks/model_benchmarks/pytorch_gpt2.py:66]: Do we add E5M2 as well?

File: superbench/benchmarks/model_benchmarks/pytorch_gpt2.py [hunk 3]
```diff
@@ -99,9 +110,31 @@ def _create_model(self, precision):
...
+            logger.error(
+                f'Create model with fp8 failed - model: {self._name}, precision: {precision},'
+                ' message: Cannot find transformer_engine.'
+            )
+            return False
+        if enable_fp8 and not self._gpu_available:
+            logger.error(
+                f'Create model with fp8 failed - model: {self._name}, precision: {precision},'
+                ' message: FP8 is only supported on GPU.'
+            )
```
Comment (cp5555) [inline superbench/benchmarks/model_benchmarks/pytorch_gpt2.py:124]: Why _gpu_available is only for FP8?

File: superbench/benchmarks/model_benchmarks/pytorch_bert.py (hunk 1 above)
Comment (abuccts) [reply superbench/benchmarks/model_benchmarks/pytorch_bert.py:193]: yes, removed

File: superbench/benchmarks/model_benchmarks/pytorch_gpt2.py (hunk 2 above)
Comment (abuccts) [reply superbench/benchmarks/model_benchmarks/pytorch_gpt2.py:66]: can only run with e4m3 (fp/bp f4m3) and hybrid (fp e4m3, bp e5m2), pure e5m2 (fp/bp both e5m2) is not supported in te

File: superbench/benchmarks/model_benchmarks/pytorch_gpt2.py (hunk 3 above)
Comment (abuccts) [reply superbench/benchmarks/model_benchmarks/pytorch_gpt2.py:124]: other precisions (fp16/fp32) can run on cpu but current te fp8 can only run on gpu, there's no cpu implementation

File: superbench/benchmarks/model_benchmarks/pytorch_gpt2.py (hunk 2 above)
Comment (cp5555) [reply superbench/benchmarks/model_benchmarks/pytorch_gpt2.py:66]: Got it

File: superbench/benchmarks/model_benchmarks/pytorch_base.py [hunk 4]
```diff
@@ -44,6 +48,40 @@ def _set_force_fp32(self):
...
+        https://github.com/huggingface/accelerate/blob/v0.17.1/src/accelerate/utils/transformer_engine.py#L24
+
+        Args:
+            model (torch.nn.Module): Torch model.
+        """
+        if not te:
+            return
+        for name, m in model.named_children():
+            if isinstance(m, torch.nn.Linear):
+                if any(p % 16 != 0 for p in m.weight.shape):
```
Comment (yzygitzh) [inline superbench/benchmarks/model_benchmarks/pytorch_base.py:66]: What does 16 mean here?

Comment (abuccts) [reply superbench/benchmarks/model_benchmarks/pytorch_base.py:66]: transformer engine fp8 op only supports 16-byte aligned

**Pull request merged**